### PR TITLE
Add CTST test for restore requeuing

### PR DESCRIPTION
**Pull request by KillianG**
Will need to release backbet and put the right version instead of using custom image

### CODE REVIEW

File: .github/scripts/end2end/configs/zenkoversion.yaml
Comment (ghost) [inline .github/scripts/end2end/configs/zenkoversion.yaml]: Leaving a note to not forget reverting this and bump backbeat instead when it's released

File: .github/scripts/end2end/run-e2e-ctst.sh
Comment (ghost) [inline .github/scripts/end2end/run-e2e-ctst.sh]: We can keep the last line here, no need to remove it

File: .github/scripts/end2end/run-e2e-ctst.sh
Comment (ghost) [inline .github/scripts/end2end/run-e2e-ctst.sh]: I would argue that this should be done in the preparation steps in the CI, not here, as running CTST should be idempotent from the cluster state

File: tests/ctst/features/azureArchive.feature
Comment (ghost) [inline tests/ctst/features/azureArchive.feature:133]: I know other tests are are also flagged as flaky, but do we know why they are flaky? can we make them more robust?

File: tests/ctst/steps/azureArchive.ts
Comment (ghost) [inline tests/ctst/steps/azureArchive.ts:8]: Do you think a work should be done on CTST to make it simpler to run other cli tools like sorbetclt here? With a set of methods, from CTST, to execute a tool with command line parameters and then handle the errors for you?

File: tests/ctst/steps/azureArchive.ts
Comment (ghost) [inline tests/ctst/steps/azureArchive.ts]: Do we really need that as an environment variable? The world parameters can be extended with this uuid (and maybe with a more meaningful name), what do you think?

File: tests/ctst/features/azureArchive.feature
Comment (KillianG) [reply tests/ctst/features/azureArchive.feature:133]: The transition to cold is flaky.. I think we can make them more robust but I thinks this better goes in a separate ticket

Though, it would be nice to set a step as flaky instead of the full test, actually only the part where we wait for the storage class to change here is flaky...
It would be clearer to have only this step as flaky

File: tests/ctst/steps/azureArchive.ts
Comment (KillianG) [reply tests/ctst/steps/azureArchive.ts:8]: Yep, as we discussed in the office on friday, I would say that an implementation of kube api would be very nice here

File: tests/ctst/steps/azureArchive.ts
Comment (francoisferrand) [inline tests/ctst/steps/azureArchive.ts]: shouldn't we have a timeout here?
or is it handled by CTST already?

File: tests/ctst/steps/azureArchive.ts
Comment (francoisferrand) [inline tests/ctst/steps/azureArchive.ts]: extra indent for readability

File: .github/scripts/end2end/run-e2e-ctst.sh
Comment (francoisferrand) [inline .github/scripts/end2end/run-e2e-ctst.sh]: `BACKBEAT_UUID` : this is the prefix of all topics, right?
→ if so, this is not `backbeat UUID`, but actually just the 'InstanceID' of zenko, as used also in mongo...

 Instead of relying on the "shared" prefix in all topics (which could probably be found in Zenko CR instead) and hard-coded extra prefixes, it would be better to simply check the `coldStorageArchiveTopicPrefix`, `coldStorageRestoreTopicPrefix` and `coldStorageStatusTopicPrefix` fields in the config : which will contain both instance id and '-cold-archive/....' prefix.

**Review by francoisferrand — APPROVED** [review]
LGTM once comments are adressed

File: tests/ctst/steps/azureArchive.ts
Comment (KillianG) [reply tests/ctst/steps/azureArchive.ts]: Yes it is already handled by CTST

File: .github/workflows/end2end.yaml
Comment (ghost) [inline .github/workflows/end2end.yaml]: As this duplicates the kafka step when installing yq, I would suggest creating a dedicated action file for that, that can be shared with `build-kafka`.

Comment (ghost) [reply .github/workflows/end2end.yaml]: You can also avoid the `cd` by setting the `working-directory` parameter to the solution folder

File: tests/ctst/steps/azureArchive.ts
Comment (ghost) [inline tests/ctst/steps/azureArchive.ts]: nit
```suggestion
        const objName = objectName || this.getSaved<string>('objectName');
```

File: tests/ctst/steps/azureArchive.ts
Comment (ghost) [inline tests/ctst/steps/azureArchive.ts]: don't we need some kind of await here, before continuing?

File: tests/ctst/world/Zenko.ts
Comment (ghost) [inline tests/ctst/world/Zenko.ts:93]: This can be set above the generic line for consistency

File: .github/workflows/end2end.yaml
Comment (KillianG) [reply .github/workflows/end2end.yaml]: Adding a specific file just for those 2 lines of installing yq feels a bit overkill to me
I now use `working-directory` though

File: tests/ctst/features/azureArchive.feature
Comment (francoisferrand) [inline tests/ctst/features/azureArchive.feature]: > blob for object "obj-1" must be rehydrated "without" notifying the queue

This does not really describe the scenario.
"without notifying" is actually not an observable outcome of this scenario, but more a detail of the implementation of the "failure".

Since the scenario is "faile restore must be retried", it should look something the following, where the scenario describes the _intent_ and code deals with the gory details of "how do we actually do/simulate this":

```
When i restore object "obj-1" for <restoreDays> days
    And object "obj-1" fails to rehydrate
    Then object "obj-1" should be "transitioning" and have the storage class "e2e-azure-archive"
    And an entry should be added to dead-letter-queue

When i run sorbetctl to retry failed restore for "e2e-azure-archive" location
    Then object "obj-1" should be "restored" and have the storage class "e2e-azure-archive"
    And object "obj-1" should expire in <restoreDays> days
    And object "obj-1" should have the same data
```

(this does not really affect the actual implementation of the steps/..., but more how to they are called)

Comment (KillianG) [reply tests/ctst/features/azureArchive.feature]: >     And object "obj-1" fails to rehydrate
I am not sure about this step, as this will just do nothing at all..

>     And an entry should be added to dead-letter-queue
This step seems a bit to "technical" to me, I think in such scenarios we don't want to know how it works internally, just the fact that it fails and it can be retried... no ?


I'm thinking of a way to make in clearer

File: tests/ctst/steps/azureArchive.ts
Comment (francoisferrand) [inline tests/ctst/steps/azureArchive.ts:392]: please introduce helper function (avoid duplicating this code!), and add comment to explain what/why this function does what it does (i.e. how it simulates a failure to rehydrate)

File: .github/scripts/end2end/configs/zenkoversion.yaml
Comment (francoisferrand) [inline .github/scripts/end2end/configs/zenkoversion.yaml]: and should not be changed (not even be in history, it uses introduces noise after the PR is merged : so remove the "change" + "revert")

**Review by KazToozs — COMMENTED** [review]
LGTM

**Review by KazToozs — APPROVED** [review]

**Pull request merged**
